src/app/qgsmaptooladdpart.cpp

```diff
@@ -226,37 +214,37 @@ void QgsMapToolAddPart::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
 
       vlayer->triggerRepaint();
 
-      if ( ( !isGeometryEmpty ) && QgsWkbTypes::isSingleType( vlayer->wkbType() ) )
-      {
-        emit messageEmitted( tr( "Add part: Feature geom is single part and you've added more than one" ), Qgis::Warning );
-      }
-
       return;
     }
 
-    case 1:
-      errorMessage = tr( "Selected feature is not multi part." );
+    case QgsGeometry::OperationResult::InvalidInputGeometryType:
+      errorMessage = tr( "New part's geometry is empty or invalid." );
       break;
 
-    case 2:
-      errorMessage = tr( "New part's geometry is not valid." );
+    case QgsGeometry::OperationResult::AddPartNotMultiGeometry:
+      errorMessage = tr( "Selected feature is not multi part." );
       break;
 
-    case 3:
-      errorMessage = tr( "New polygon ring not disjoint with existing polygons." );
+    case QgsGeometry::OperationResult::SelectionIsEmpty:
+      errorMessage = tr( "No feature selected. Please select a feature with the selection tool or in the attribute table." );
       break;
 
-    case 4:
-      errorMessage = tr( "No feature selected. Please select a feature with the selection tool or in the attribute table" );
+    case QgsGeometry::OperationResult::SelectionIsGreaterThanOne:
+      errorMessage = tr( "Several features are selected. Please select only one feature to which an island should be added." );
       break;
 
-    case 5:
-      errorMessage = tr( "Several features are selected. Please select only one feature to which an island should be added." );
+    case QgsGeometry::OperationResult::AddPartSelectedGeometryNotFound:
+      errorMessage = tr( "Selected geometry could not be found." );
       break;
 
-    case 6:
-      errorMessage = tr( "Selected geometry could not be found" );
+    case QgsGeometry::OperationResult::InvalidBaseGeometry:
+      errorMessage = tr( "Base geometry is not valid." );
       break;
+
+    default:
+      // Should not reach here
+      // Other OperationResults should not be returned by addPart
+      errorMessage = tr( "Unexpected OperationResult: %1" ).arg( errorCode );
```

Comment:
would a `Q_ASSERT` be justified here?

Reply:
What's to gain? It is already within the 'handle errors' part of the code, any return type other than *Success* is handled the same, only the message to the user changes.

Reply:
If a new enum value is added in the future, nobody will get notified because of the `default` catch-all. With a `Q_ASSERT` it's more likely a developer will catch this with a use case at hand and introduce an error message.

Reply:
I'd rather just see the "default" clause removed entirely, so we get a compile time warning.

Reply:
@nyalldawson We can't do that, there are other `OperationResult`s completely unrelated to adding parts to geometries. Removing the default will cause a compiler warning right away!

@m-kuhn The `default` will show the generic `Unexpected OperationResult: %1` message so it will get noticed, at least later if not sooner! I would not like my debug build crashing for such a reason that is not operation critical.

Reply:
In that case you add all the existing other values explicility instead of the default clause -- you'll still get a compile warning if someone adds a new value and the responsibility lies with them to correctly handle it in this block

Reply:
I can't say I like this approach. It means one will have to update this tool, and add practically unreachable code, whenever a completely unrelated OperationResult is added.

Reply:
That's exactly the intention, and it's why it's considered good practice (see C++ core guidelines)